File: frontend/lib/ai-providers.ts
import { api } from "./api-client";

export interface AiProviderKey {
  id: string;
  organization_id: string | null;
  provider_name: "gemini" | "openai" | "claude" | "deepseek" | "openrouter" | string;
  label: string;
  masked_api_key: string;
  model_name: string;
  priority: number;
  is_active: boolean;
  last_error_at: string | null;
  error_count: number;
  last_used_at: string | null;
  usage_count: number;
  status: "active" | "rate_limited" | "warning" | "inactive" | string;
  created_at: string;
  updated_at: string;
}

export interface AiProviderKeyCreatePayload {
  provider_name: string;
  label: string;
  api_key: string;
  model_name: string;
  priority: number;
  is_active?: boolean;
}

export interface AiProviderKeyUpdatePayload {
  label?: string;
  api_key?: string;
  model_name?: string;
  priority?: number;
  is_active?: boolean;
}

export interface AiProviderKeyTestResult {
  status: string;
  provider: string;
  model: string;
  latency_ms: number;
  response_sample: string;
}

export async function getAiProviderKeys(): Promise<AiProviderKey[]> {
  const res = await api.get<AiProviderKey[]>("/ai-providers");
  return res || [];
}

export async function createAiProviderKey(
  payload: AiProviderKeyCreatePayload
): Promise<AiProviderKey> {
  return await api.post<AiProviderKey>("/ai-providers", payload);
}

export async function updateAiProviderKey(
  id: string,
  payload: AiProviderKeyUpdatePayload
): Promise<AiProviderKey> {
  return await api.put<AiProviderKey>(`/ai-providers/${id}`, payload);
}

export async function deleteAiProviderKey(id: string): Promise<{ data: { deleted: boolean; id: string } }> {
  return await api.delete<{ data: { deleted: boolean; id: string } }>(`/ai-providers/${id}`);
}

export async function testStoredAiKey(id: string): Promise<AiProviderKeyTestResult> {
  return await api.post<AiProviderKeyTestResult>(`/ai-providers/${id}/test`);
}

export async function testRawAiKey(payload: {
  provider_name: string;
  api_key: string;
  model_name: string;
}): Promise<AiProviderKeyTestResult> {
  return await api.post<AiProviderKeyTestResult>("/ai-providers/test-raw", payload);
}

File: backend/app/schemas/ai_providers.py
from datetime import datetime
from uuid import UUID
from pydantic import BaseModel, Field, ConfigDict


class AiProviderKeyCreate(BaseModel):
    provider_name: str = Field(default="gemini", description="gemini, openai, claude, deepseek, openrouter")
    label: str = Field(min_length=1, max_length=100, description="Display name for this key")
    api_key: str = Field(min_length=5, description="Plaintext API key to be encrypted and stored")
    model_name: str = Field(default="gemini-3.6-flash", max_length=100)
    priority: int = Field(default=1, ge=1, le=100, description="Lower number = used first")
    is_active: bool = True


class AiProviderKeyUpdate(BaseModel):
    label: str | None = None
    api_key: str | None = None
    model_name: str | None = None
    priority: int | None = Field(default=None, ge=1, le=100)
    is_active: bool | None = None


class AiProviderKeyRead(BaseModel):
    id: UUID
    organization_id: UUID | None
    provider_name: str
    label: str
    masked_api_key: str
    model_name: str
    priority: int
    is_active: bool
    last_error_at: datetime | None = None
    error_count: int = 0
    last_used_at: datetime | None = None
    usage_count: int = 0
    status: str = "active"  # active, rate_limited, error
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class AiProviderKeyTestRequest(BaseModel):
    provider_name: str = "gemini"
    api_key: str
    model_name: str = "gemini-3.6-flash"


class AiProviderKeyTestResult(BaseModel):
    status: str
    provider: str
    model: str
    latency_ms: int
    response_sample: str
